src/adapter/src/coord/ddl.rs

```diff
@@ -733,29 +739,46 @@ impl Coordinator {
             })),
             SINCE_GRANULARITY,
         );
+        self.update_retention(policy, |item| item.is_retained_metrics_object())
+    }
+
+    fn update_default_retention(&mut self) {
+        let lag = self.catalog().system_config().default_retention_timestamp();
+        let policy = ReadPolicy::lag_writes_by(lag, SINCE_GRANULARITY);
+        self.update_retention(policy, |item| item.uses_default_retention())
+    }
+
+    fn update_retention(
+        &mut self,
+        policy: ReadPolicy<Timestamp>,
+        update_item: impl Fn(&CatalogItem) -> bool,
+    ) {
         let storage_policies = self
-            .catalog()
-            .entries()
-            .filter(|entry| {
-                entry.item().is_retained_metrics_object()
-                    && entry.item().is_compute_object_on_cluster().is_none()
+            .storage_read_capabilities
+            .iter()
+            .filter_map(|(id, _)| {
+                self.catalog()
+                    .try_get_entry(id)
+                    .map(|entry| (*id, entry.item()))
             })
-            .map(|entry| (entry.id(), policy.clone()))
+            .filter_map(|(id, item)| update_item(item).then(|| (id, policy.clone())))
             .collect::<Vec<_>>();
+
         let compute_policies = self
-            .catalog()
-            .entries()
-            .filter_map(|entry| {
-                if let (true, Some(cluster_id)) = (
-                    entry.item().is_retained_metrics_object(),
-                    entry.item().is_compute_object_on_cluster(),
-                ) {
-                    Some((cluster_id, entry.id(), policy.clone()))
-                } else {
-                    None
-                }
+            .compute_read_capabilities
+            .iter()
+            .filter_map(|(id, _)| {
+                self.catalog()
+                    .try_get_entry(id)
+                    .map(|entry| (*id, entry.item()))
+            })
+            .filter_map(|(id, item)| {
+                item.cluster_id().and_then(|cluster_id| {
+                    update_item(item).then(|| (cluster_id, id, policy.clone()))
+                })
```

Comment:
Interesting that this fixed the errors. That leaves me with two questions:

1. What are these objects that don't exist?
2. Why aren't they causing errors when we update our read holds every second?

Reply:
Fixed by implementing coord checker: everything in {compute,storage}_read_capabilities but transient compute should have a corresponding catalog entry